cpp/ql/lib/semmle/code/cpp/ir/dataflow/MustFlow.qll

```diff
@@ -198,12 +197,13 @@ private module Cached {
   }
 
   cached
-  predicate step(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
-    instructionToOperandStep(nodeFrom.asInstruction(), nodeTo.asOperand())
-    or
-    flowThroughCallable(nodeFrom.asInstruction(), nodeTo.asInstruction())
+  predicate step(Instruction nodeFrom, Instruction nodeTo) {
+    exists(Operand mid |
+      instructionToOperandStep(nodeFrom, mid) and
+      operandToInstructionStep(mid, nodeTo)
```

Comment:
Why is this now needed, while it wasn't before?

Reply:
The `step` function must be a binary relation between two columns of the same type (otherwise a transitive closure over the step relation doesn't make any sense). So if it has to be a binary relation between instructions, then we need to do two things in order to step from one instruction to another:
1. An `Instruction` to `Operand` step.
2. An `Operand` to `Instruction` step.

Previously, since the columns were `DataFlow::Node`s (which contains both instructions and operands), this was done in two iterations of the `step` predicate (i.e., the `instructionToOperandStep(nodeFrom.asInstruction(), nodeTo.asOperand())` disjunct and `operandToInstructionStep(nodeFrom.asOperand(), nodeTo.asInstruction())` disjunct), but we can't do that now since the columns are both instructions. So instead we do two "old" steps in one "new" step.